### DISCUSSION TITLE
4장 4.1.9 스레드 풀에 대해

### DISCUSSION THREAD

**Discussion by jjuyeon** [concordians/real-mysql#16] [Q&A] [answered]
스레드 풀 시스템 변수의 적용 범위는 ‘각 엔진’일지 ‘서버 전체’일지 궁금합니다.

**Comment by sungyujeon** [comment, marked as answer]
우선은.. 공식 docs 내 [thread-pool](https://dev.mysql.com/doc/refman/8.0/en/faqs-thread-pool.html)에 대한 내용을 보면, 우리가 흔히 알고 있는 connection pool 과는 조금 다른 개념인 것 같습니다. 

```txt
How is the Thread Pool different from the client side Connection Pool?

(생략)
By contrast, the Thread Pool operates on the MySQL server side 
and is designed to manage the execution of inbound concurrent connections 
and queries as they are received from the client connections 
accessing the back-end MySQL database. 
Because of the separation of duties, the MySQL Connection Pool and Thread Pool 
are orthogonal and can be used independent of each other.
```

스레드풀은 '오버헤드를 줄이고 성능을 향상시기 위해 사용자 요청에 대한 스레드를 어떻게 handling할 것인가?'에 대한 해결책으로서 나온 기술입니다.
서버 사이드에서 작동하고, 요청이 들어온 client connections와 쿼리 실행을 관리합니다. 
이러한 관리는 MySQL 서버(MySQL 엔진, 스토리지 엔진 등..) 전반에 걸쳐 영향을 미칠 것입니다.

thread-pool과 관련된 [시스템 변수](https://dev.mysql.com/doc/refman/8.0/en/thread-pool-operation.html)에서도 보면, 예를 들어 [thread_pool_max_transactions_limit](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_thread_pool_max_transactions_limit) 설정 후 수용 가능한 transaction 수에 limit이 있을 경우, 이를 초과한 connection은 대기해야 하거나 끊기게 됩니다. 해당 connection을 제어하는 것은 아마도 스토리지 엔진이기 보다는 MySQL 엔진에 해당하는 커넥션 핸들러이지 않을까 싶습니다.

반면 앞선 [FAQ](https://dev.mysql.com/doc/refman/8.0/en/faqs-thread-pool.html)와 [thread_pool_high_priority_connection](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_thread_pool_high_priority_connection) 시스템 변수를 보면, 
```txt
When should I use the Thread Pool?

The MySQL Threads_running variable keeps track of the number of concurrent statements 
currently executing in the MySQL Server. 

If you are using the innodb_thread_concurrency to limit the number of concurrently executing statements, 
you should find that the Thread Pool solves the same problem, only better, 
by assigning connections to thread groups, 
then queuing executions based on transactional content, user defined designations, and so forth.
```
thread pool과 관련된 설정이 innodb 스토리지 엔진의 statement 처리 우선순위에도 영향을 미친다고 추측할 수 있을 것 같습니다.

결론적으로 시스템 변수에 따라 각 엔진에 영향을 미칠 수도, 서버 전체에 영향을 미칠 수도 있지 않을까 싶습니다.

---

논외로, 시스템 변수가 서버 전체에 적용될지, 각 세션에 적용될지를 물어보는 것은 아니였을까 해서 혹시나 하여 추가합니다.

[2장 설치와 설정의 서버 설정](https://github.com/concordians/real-mysql/blob/main/chapter02/02_%EC%84%A4%EC%B9%98%EC%99%80%EC%84%A4%EC%A0%95_%EC%A0%84%EC%84%A0%EA%B7%9C.md#%EC%8B%9C%EC%8A%A4%ED%85%9C-%EB%B3%80%EC%88%98%EC%9D%98-%ED%8A%B9%EC%A7%95)에서도 나왔듯이, 시스템 변수 자체는 서버 실행 시 옵션을 통해서, 또는 설정 파일(우선순위에 따르는)로 선언 및 할당할 수 있습니다. 시스템 변수는 다시 글로벌 변수와 세션 변수로 나뉩니다. 글로벌 변수는 단일 MySQL 서버 인스턴스에 적용되는 옵션이고, 세션 변수는 MySQL 각 클라이언트가 부여받는 옵션입니다.

thread pool 관련 시스템 변수를 보면,
[thread_pool_size](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_thread_pool_size)는 Scope이 Global이고,
[thread_pool_high_priority_connection](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_thread_pool_high_priority_connection)은 Scope이 Global & Session 입니다.

즉, 각 시스템변수의 scope별로 전체 서버 인스턴스에 적용될 수도, 세션에 적용될 수도 있습니다.
